plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -194,6 +221,61 @@ public IcebergTableHandle getTableHandle(ConnectorSession session, SchemaTableNa
                 Optional.ofNullable(nameMappingJson));
     }
 
+    @Override
+    public boolean isSupportedVersionType(ConnectorSession session, SchemaTableName tableName, PointerType pointerType, io.trino.spi.type.Type versioning)
+    {
+        switch (pointerType) {
+            case TEMPORAL:
+                return versioning instanceof TimestampWithTimeZoneType || versioning instanceof TimestampType;
+            case TARGET_ID:
+                return versioning == BIGINT;
+        }
+        throw new TrinoException(NOT_SUPPORTED, "Version type " + versioning + " is not supported in Iceberg");
+    }
+
+    private Optional<Long> getSnapshotIdFromVersion(Table table, Optional<ConnectorTableVersion> versionOptional)
+    {
+        Optional<Long> snapshotId = Optional.empty();
+        if (versionOptional.isPresent()) {
+            ConnectorTableVersion version = versionOptional.get();
+            io.trino.spi.type.Type versionType = version.getVersionType();
+
+            switch (version.getPointerType()) {
+                case TEMPORAL:
+                    verify(versionType instanceof TimestampWithTimeZoneType || versionType instanceof TimestampType,
+                            "Iceberg temporal time travel only supports types TimestampWithTimeZone and Timestamp, but got " + versionType);
+
+                    long epochMillis;
+                    if (versionType instanceof TimestampWithTimeZoneType) {
+                        if (((TimestampWithTimeZoneType) versionType).isShort()) {
+                            epochMillis = unpackMillisUtc((long) version.getVersion());
+                        }
+                        else {
+                            LongTimestampWithTimeZone timestamp = (LongTimestampWithTimeZone) version.getVersion();
+                            epochMillis = timestamp.getEpochMillis();
+                        }
+                    }
+                    else {
+                        if (((TimestampType) versionType).isShort()) {
+                            epochMillis = ((long) version.getVersion()) / 1_000;
```

Comment:
There's a `MICROSECONDS_PER_MILLISECOND` constant, which might make this a little clearer

Reply:
updated